NOTES.md

```diff
@@ -0,0 +1,60 @@
+# Notes from review
+
+* WebKit's documentation and intro on parking lot and its paradigm is quite good
+
+* `cargo test` fails with compilation errors on Linux
+  * had an ancient lock file on my system and `cargo update` was needed
+
+* Various bits and pieces of the library are centered around compatibility with
+  various versions of Rust. This naturally isn't needed when integrating into
+  libstd, but it can often reduce the readability of code and/or make it more
+  difficult to follow. Do we want to have a clear version policy for parking_lot
+  on crates.io which is geared towards simplifying the code?
+
+* Overall quite worried about atomic orderings. Everything should be `SeqCst`
+  until proven otherwise, and even then needs quite a lot of proof (IMO) to not
+  use `SeqCst`. Being fast-and-loose with orderings seems like we're trying to
+  be too clever by half.
```

Comment:
I disagree. In this case the orderings are very clear: `Acquire` when acquiring a lock and `Release` when releasing a lock. All other operations have no ordering requirements, and can use `Relaxed`.

Reply:
Sorry but this isn't gonna fly with me at least. Atomic orderings are notoriously hard to get right, and I think it's just wrong that "hey it's a lock so acquire/release and everything that looks unrelated is relaxed".

I've personally got two requirements for atomic orderings:

* Either everything is `SeqCst`
* Or there's documentation lengthily explaining why `SeqCst` isn't used. This involves concrete benchmarks showing the performance different, pointers to other code in the world which is also thoroughly vetted and using more relaxed atomics, etc. 

There's basically no way that this can be maintained to the level of quality expected from the standard library if our attitude is "no documentation necessary, and if it looks like a lock acquire/release otherwise relaxed everything".

Reply:
I agree that atomic algorithms need to be carefully documented to argue for their correctness. With `Relaxed`, I agree that it should explain why there are no ordering requirements, and why it is okay not to synchronize this access.

However, I disagree that using `SeqCst` makes the argument any simpler. `SeqCst` works fine as a "I don't want to think about this now so let's use the least-likely-to-be-incorrect option". So, it makes things simpler if you *don't* have a careful argument. But reasoning *based on* `SeqCst` basically requires reasoning by exhaustively looking at all interleavings, and that is certainly not maintainable. Hence I find any remaining `SeqCst` to be a strong sign that something *really strange* is going on that *needs* sequential consistency (which is extremely rare).

After all, acquire/release operations are literally called that way because they are exactly what one needs to acquire/release a lock.

Reply:
At the absolute bare minimum anything not `SeqCst` needs to be documented in my opinion. It's a good point though that I personally assume `SeqCst` isn't "I thought about this and need it" but rather "I didn't think about this so this is the default I chose". I think it's safe to assume that in the sense that if an ordering isn't documented and it's `SeqCst` we can assume it's just "didn't think about this will relax later if necessary".

I am not personally comfortable enough with atomic orderings to sign off on this if the non-`SeqCst` orderings are documented and remain. I'd want review from someone who is comfortable reviewing the orderings.

Reply:
I can help check the orderings. And it seems we are in agreement that more docs are needed either way, that should hopefully make this easier. The docs should explain either why no synchronization is needed, or which kind of synchronization is happening.  So e.g. the acquire load on a lock-acquire method should say that this synchronizes-with the release store of the lock-release operation of whoever previously held the lock, thus establishing the necessary synchronization between the previous thread that held the lock and the new thread that now got it.

Reply:
I disagree about `SeqCst` by default. If I see `SeqCst` my immediate reaction is "the programmer likely has no idea about atomics, and this code is full of bugs". Atomics, even when "sequentially consistent" guarantee much less than most programmęrs would expect, and actually needing "sequential consistency" guarantee is very rare to the point where I've never seen it being necessary (albeit there may be usecases where it may need to be used).

`std` uses `SeqCst` pretty much everywhere in its `mpsc` implementation. The implementation is pretty much broken as evidenced by issues like https://github.com/rust-lang/rust/issues/39364.

Reply:
FWIW, I agree with the sentiment that release-acquire should be the default and anything else require extra comments -- but that is not a consensus in the Rust teams, see https://github.com/rust-lang/rfcs/pull/2503 for some discussion.